src/Host/Client/Impl/Session/RSession.cs

```diff
@@ -325,8 +325,7 @@ public Task StartHostAsync(RHostStartupInfo startupInfo, IRSessionCallback callb
             // Try graceful shutdown with q() first.
             if (host != null) {
                 try {
-                    host.QuitAsync().SilenceException<Exception>().DoNotWait();
-                    await Task.WhenAny(hostRunTask, Task.Delay(2000)).Unwrap();
+                    host.QuitAsync().SilenceException<Exception>().Wait(10000);
```

Comment:
Waits for task completion or 10 seconds, whichever comes first. Yes, it blocks. If it doesnt block then VS will proceed with shutdown. Here we are in 'IOleComponent.Terminate` call path